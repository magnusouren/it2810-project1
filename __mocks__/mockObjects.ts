import { Drink } from '../src/types';

export const defaultDrinkResponse = {
  idDrink: '11118',
  strDrink: 'Super Drink',
  ingredients: [],
  strGlass: 'Highball',
  strInstructions: 'Instructions for drink',
  strDrinkThumb: 'https://localhost:3000/drink/11118.jpg',
  strCategory: 'Category',
  strAlcoholic: 'Alcoholic',
  strIngredient1: 'Ingredient 1',
  strMeasure1: 'Measure 1',
  strIngredient2: 'Ingredient 2',
  strMeasure2: 'Measure 2',
};

export const mockDrink: Drink = {
  idDrink: '11118',
  strDrink: 'Super Drink',
  ingredients: [
    { ingredient: 'Ingredient 1', measure: 'Measure 1' },
    { ingredient: 'Ingredient 2', measure: 'Measure 2' },
  ],
  strGlass: 'Highball',
  strInstructions: 'Instructions for drink',
  strDrinkThumb: 'https://localhost:3000/drink/11118.jpg',
  strCategory: 'Category',
  strAlcoholic: 'Alcoholic',
};

export const alternativeDrink: Drink = {
  idDrink: '22222',
  strDrink: 'Foo bar',
  ingredients: [],
  strGlass: 'Foo',
  strInstructions: 'Instructions for foo drink',
  strDrinkThumb: 'https://localhost:3000/drink/2.jpg',
  strCategory: 'Category',
  strAlcoholic: 'Alcoholic',
};

export const randomDrink = {
  idDrink: '9118',
  strDrink: 'Random Drink',
  strDrinkThumb: 'https://localhost:3000/drink/11118.jpg',
  strCategory: 'Category',
  strGlass: 'Highball',
  strAlcoholic: 'Alcoholic',
};
