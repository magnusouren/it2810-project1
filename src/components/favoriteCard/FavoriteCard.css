.favorites {
  margin: auto;
  width: 60%;
}

@media (width <= 992px) {
  .favorites {
    width: 90%;
  }
}

.favorites ul {
  list-style: none;
  padding: 0;
  width: 100%;
}

.favorites li {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin: 2rem 0;
  padding-bottom: 2rem;
  border-bottom: solid 1px #ccc;
}

.favorites li:last-child {
  border-bottom: none;
}

.favorites img {
  width: 100px;
  height: 100px;
  border-radius: 0.5rem;
  margin-right: 2rem;
}

.favorites h3 {
  font-size: 1.5rem;
  margin-bottom: 0.5rem;
  color: #313030;
  text-align: center;
  font-weight: 500;
}

.favorites a {
  text-decoration: none;
  color: #313030;
}

.favorites h3:hover {
  text-decoration-line: underline;
  text-decoration-color: #f2c2a6;
}

.favorites p {
  font-weight: 300;
}

.remove-favorite {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;
}

.remove-favorite .material-symbols-outlined {
  font-size: 2rem;
}

.remove-favorite .material-symbols-outlined:hover {
  cursor: pointer;
  color: #606060;
}

.extra-info {
  color: gray;
  text-align: center;
}
