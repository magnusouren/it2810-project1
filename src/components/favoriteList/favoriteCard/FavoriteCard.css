.favorites {
  margin: auto;
  width: 60%;
}

.favorites ul {
  list-style: none;
  padding: 0;
  width: 100%;
}

.favorites li {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin: 2rem 0;
  padding-bottom: 2rem;
  border-bottom: solid 1px #ccc;
}

.favorites li:last-child {
  border-bottom: none;
}

.favorites img {
  width: 100px;
  height: 100px;
  border-radius: 0.5rem;
  margin-right: 2rem;
}

.favorites h3 {
  font-size: 1.5rem;
  margin-bottom: 0.5rem;
  color: #313030;
}

.favorites h3:hover {
  text-decoration: underline #f2c2a6;
}

.material-symbols-outlined {
  font-size: 2rem;
}

.material-symbols-outlined:hover {
  cursor: pointer;
  color: #606060;
}

.remove-favorite {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;
}

.extra-info {
  color: gray;
  text-align: center;
}
