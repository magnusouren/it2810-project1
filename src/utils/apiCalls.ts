import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
