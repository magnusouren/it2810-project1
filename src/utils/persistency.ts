import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
