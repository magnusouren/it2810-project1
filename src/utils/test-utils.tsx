import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';

import { DrinkCard } from '../components/drink/DrinkCard';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );

export const renderWithRouterQueryClientAndDrinkId = (id: string) =>
  render(
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>,
  );
