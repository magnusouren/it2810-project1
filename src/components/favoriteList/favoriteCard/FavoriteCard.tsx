import './FavoriteCard.css';

import { FC } from 'react';
import { Link } from 'react-router-dom';

import { Drink } from '../../../types';
import { removeFavorite } from '../../../utils/persistency';

interface FavoriteCardProps {
  drink: Drink;
}

export const FavoriteCard: FC<FavoriteCardProps> = ({ drink }) => {
  const handleClick = () => {
    removeFavorite(drink.idDrink);
    window.location.reload();
  };
  return (
    <li key={drink.strDrink}>
      <img src={drink.strDrinkThumb} alt={drink.strDrink} />
      <div>
        <Link to={`/drink/${drink.idDrink}`}>
          <h3>{drink.strDrink}</h3>
        </Link>
        <p className='extra-info'>
          {drink.strCategory} - {drink.strGlass}
        </p>
      </div>
      <button onClick={handleClick} className='remove-favorite'>
        <span className='material-symbols-outlined'>heart_minus</span>
      </button>
    </li>
  );
};
