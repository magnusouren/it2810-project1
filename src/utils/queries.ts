import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
