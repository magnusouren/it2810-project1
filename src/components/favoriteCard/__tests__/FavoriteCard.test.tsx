import { fireEvent } from '@testing-library/react';
import { vi } from 'vitest';

import { mockDrink } from '../../../../__mocks__/mockObjects';
import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { FavoriteCard } from '../FavoriteCard';

describe('FavoriteCard', () => {
  it('should render correctly', () => {
    const { container } = renderWithRouterAndQueryClient(
      <FavoriteCard drink={mockDrink} handleRemoveFavorite={vi.fn()} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should show correct category and glass', async () => {
    const { getByText } = renderWithRouterAndQueryClient(
      <FavoriteCard drink={mockDrink} handleRemoveFavorite={vi.fn()} />,
    );
    expect(getByText(`${mockDrink.strCategory} - ${mockDrink.strGlass}`)).toBeDefined();
  });

  it('should show correct name of drink', async () => {
    const { getByText } = renderWithRouterAndQueryClient(
      <FavoriteCard drink={mockDrink} handleRemoveFavorite={vi.fn()} />,
    );
    expect(getByText(mockDrink.strDrink)).toBeDefined();
  });

  it('should show correct image of drink', async () => {
    const { getByAltText } = renderWithRouterAndQueryClient(
      <FavoriteCard drink={mockDrink} handleRemoveFavorite={vi.fn()} />,
    );
    expect(getByAltText(mockDrink.strDrink)).toBeDefined();
  });

  it('should navigate to correct drink on click', async () => {
    const { getByText } = renderWithRouterAndQueryClient(
      <FavoriteCard drink={mockDrink} handleRemoveFavorite={vi.fn()} />,
    );
    fireEvent.click(getByText(mockDrink.strDrink));
    expect(window.location.pathname).toBe('/drink/11118');
  });

  it('should contain correct heading', async () => {
    const { getByRole } = renderWithRouterAndQueryClient(
      <FavoriteCard drink={mockDrink} handleRemoveFavorite={vi.fn()} />,
    );
    expect(getByRole('heading', { name: mockDrink.strDrink })).toBeDefined();
  });

  it('should call onRemoveFavorite when "hearth-minus" is clicked', async () => {
    const mock = vi.fn();
    const { getByText } = renderWithRouterAndQueryClient(
      <FavoriteCard drink={mockDrink} handleRemoveFavorite={mock} />,
    );
    getByText('heart_minus').click();
    expect(mock).toHaveBeenCalledWith(mockDrink.idDrink);
  });
});
