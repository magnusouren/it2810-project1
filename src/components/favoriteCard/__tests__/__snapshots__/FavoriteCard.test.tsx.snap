// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoriteCard > should render correctly 1`] = `
<div>
  <li>
    <img
      alt="Super Drink"
      src="https://localhost:3000/drink/11118.jpg"
    />
    <div>
      <a
        href="/drink/11118"
      >
        <h3>
          Super Drink
        </h3>
      </a>
      <p
        class="extra-info"
      >
        Category
         - 
        Highball
      </p>
    </div>
    <button
      class="remove-favorite"
    >
      <span
        class="material-symbols-outlined"
      >
        heart_minus
      </span>
    </button>
  </li>
</div>
`;
